manpages/kstats.1
.TH KSTATS 1 "December 2018" "Version 1.5.2"

.SH NAME
kstats - show statistical FMS algorithm votes for an ivs dump and a specified WEP key
.SH SYNOPSIS
.B kstats
<ivs file> <104-bit key>
.SH DESCRIPTION
.BI kstats
is a tool designed to show the FMS algorithm votes for an ivs dump (initialization vectors) with a specified WEP key. The ivs dump can be get by using the combination of both airodump(1) and ivstools(1).
.SH EXAMPLE
.B kstats
kstats out.ivs 123456789ABCDEF123456789AB
.SH AUTHOR
This manual page was written by Adam Cecile <gandalf@le-vert.net> for the Debian system (but may be used by others).
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or any later version published by the Free Software Foundation
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.
.SH SEE ALSO
.br
.B airbase-ng(8)
.br
.B aireplay-ng(8)
.br
.B airmon-ng(8)
.br
.B airodump-ng(8)
.br
.B airodump-ng-oui-update(8)
.br
.B airserv-ng(8)
.br
.B airtun-ng(8)
.br
.B besside-ng(8)
.br
.B easside-ng(8)
.br
.B tkiptun-ng(8)
.br
.B wesside-ng(8)
.br
.B aircrack-ng(1)
.br
.B airdecap-ng(1)
.br
.B airdecloak-ng(1)
.br
.B airolib-ng(1)
.br
.B besside-ng-crawler(1)
.br
.B buddy-ng(1)
.br
.B ivstools(1)
.br
.B makeivs-ng(1)
.br
.B packetforge-ng(1)
.br
.B wpaclean(1)
.br
.B airventriloquist(8)


CHANGELOG
Changelog for 2.0.9
-fixed client and clientless attacks chopchop and frag packet injection not generating IVs because -h MAC was not set
-force frag, chopchop and aircrack-ng xterms to stay on screen after successfull command so you can see results
-cleaned the code some more
-cleaning functions needs more tweaks
-atheros users need to use options 9 (monitor mode) before other commands to create-destroys correct VAPS
-fixed "unary operator expected" msg (harmless bug) (not fully fixed because of "" generating errors in logic tests)
-tweaked 2 steps attacks (frag, chopchop) to go to menu to inject after successfull first step
-added option to jump to injection menu (added this to the main menu)
-fixed AP encryption detection for cracking step (bug was introduced by "")
-lowered adapter speed to 1M for wesside-ng attacks
-added -hold to mdk attack to allow users to see results before the windows closes

Changelog for 2.0.8 (27/01/2008)
-added wesside-ng airoscript style usage
-added interface type detection
-tweaked menu
-first attempt at making airoscript work with different adapters (add more card support)(BIG THX TO GREEN-FREQ!!!)
-atheros device are now supported
-added mdk3
-added back to main menu options
-fixed chopchop attack (BIG THX TO GREEN-FREQ!!!)
-tweaked FAKE_MAC function for atheros

Changelog for 2.0.7 BETA1 (21/01/2007)
-changed capture from IVs to full date (ivs to cap)
-updated script for new implementation of fragmentation attack
-other minor modifications and tweaks

Changelog for PRE-1.8.2-BETA1 (30/11/2006)
-fixed all commands leading to : "head: Binary: invalid number of lines" (MAJOR THX TO ANDREA)
-fixed all commands where SSID was truncated (MAJOR THX TO PILOTSNIPES)

Changelog for 1.8.1 (03/11/2006)
-bug fix release
-fixed some problem in the SSID detection function (MAJOR THX TO andreahaku)
-cleaned a few more typos and bugs
-documentation completed a bit

Changelog for 1.8 (17/10/2006)
-dropped support for aircrack-ng 0.6.1 and older
-will only work with patched version of aircrack-ng 0.6.2 (patches can be found in the forum under the programming section)
-specifique channels hopping (for example channel 1,6,11 or 1-3,5-7,9,11)
-Access point filtering during attack (airodump will only list attacked Access point)
-new SSID detection, works with ALL ssid (supports all character and spacing possible)
-new Access point selection menu that list all AP with encryption type (eliminates the WEP? problem)
-option 2) now get you directly to a full list of AP
-chopchop attack and fragmentation attack implement
-WPA dictionnary attack implemented
-changed menu 3,4,5 based on detection of AP encryption
-added option to jump to associated client list from ap selection
-color change support for xterm
-Windows size and placement fully changeable from variables settings
-changed filenames from SSID to MAC because of possible issues if weird SSID (to eliminate security menace (if SSID is something like "rm -rf /*" etc...)
-added option an option to run "1) scan 2) select 3) attack" one after the other
-added some comments in the script to help users
-added greetings message (can be disabled setting WELCOME="1" to "0")

Changelog for 1.7RC7 (24/09/2006)
-added some resilience to the configure function now works properly with hidden ssid's and rausb devices 

Changelog for 1.7RC6 (19/09/2006)
-Code cleanup
-added deauth command to solointeractiveattack function
-changed base variables (fudge factor & inject rate)

Changelog for 1.7RC5 (14/09/2006)
-Fixed minor cosmetic issue caused by carriage return char when ssid is missing (15/9/06)
-Fixed the weird character match issue!! edit at will!!!
-added an abitrary variable ($acv) to the aircrackversioncheck function which can be used in future to keep the script backwards compatible (had to guess at what version 0.6.0 of aircrack-ng displays when run on it's own? and version 0.5.so might not work??)
-used new variable in crack function to check version and run correct crack command for the aricrack version installed fixing the bug on the B|T forum.

Changelog for 1.7RC4 (13/09/2006)
-corrected minor speellling issues ;o) 
-added aircrack-ng version detection function currently all it does is display aircrack version in the menu function (should be useful when new functionality is added to aircrack to allow backwards compatibility)
-added solointeractiveattack function (designed to run just the airreplay interactive attack function only for use with an attack that has stalled i.e. data has stopped increasing, the user can try this for another injection thread! (might not actually work?? needs more testing does spawn and seems to work!)
-added option to manually add ssid if a blank ssid is detected (this has some problems when certain editors are used as this is not strictly blank but a non printable character recommend kwrite on Back|track to edit script)


src/kstats.c
/*
 *  Kstat: displays the votes of the korek attack for each keybyte
 *
 *  Copyright (C) 2006-2018 Thomas d'Otreppe <tdotreppe@aircrack-ng.org>
 *  Copyright (C) 2004, 2005  Christophe Devine
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include "aircrack-util/common.h"

#define IVBUF_SIZE (5UL * 0xFFFFFF)
#define N_ATTACKS 17

enum KoreK_attacks
{
	A_u15, /* semi-stable  15%             */
	A_s13, /* stable       13%             */
	A_u13_1, /* unstable     13%             */
	A_u13_2, /* unstable ?   13%             */
	A_u13_3, /* unstable ?   13%             */
	A_s5_1, /* standard      5% (~FMS)      */
	A_s5_2, /* other stable  5%             */
	A_s5_3, /* other stable  5%             */
	A_u5_1, /* unstable      5% no good ?   */
	A_u5_2, /* unstable      5%             */
	A_u5_3, /* unstable      5% no good     */
	A_u5_4, /* unstable      5%             */
	A_s3, /* stable        3%             */
	A_4_s13, /* stable       13% on q = 4    */
	A_4_u5_1, /* unstable      5% on q = 4    */
	A_4_u5_2, /* unstable      5% on q = 4    */
	A_neg /* helps reject false positives */
};

int K_COEFF[N_ATTACKS]
	= {15, 13, 12, 12, 12, 5, 5, 5, 3, 4, 3, 4, 3, 13, 4, 4, -20};

static void calc_votes(unsigned char * ivbuf,
					   long nb_ivs,
					   unsigned char * key,
					   int B,
					   int votes[N_ATTACKS][256])
{
	int i, j;
	long xv;
	unsigned char R[256], jj[256];
	unsigned char S[256], Si[256];
	unsigned char K[64];

	unsigned char io1, o1, io2, o2;
	unsigned char Sq, dq, Kq, jq, q;
	unsigned char S1, S2, J2, t2;

	for (i = 0; i < 256; i++) R[i] = i;

	q = 3 + B;

	memcpy(K + 3, key, B);
	memset(votes, 0, sizeof(int) * N_ATTACKS * 256);

	for (xv = 0; xv < nb_ivs; xv += 5)
	{
		memcpy(K, &ivbuf[xv], 3);
		memcpy(S, R, 256);
		memcpy(Si, R, 256);

		for (i = j = 0; i < q; i++)
		{
			jj[i] = j = (j + S[i] + K[i & 15]) & 0xFF;
			SWAP(S[i], S[j]);
		}

		i = q;
		do
		{
			i--;
			SWAP(Si[i], Si[jj[i]]);
		} while (i != 0);

		o1 = ivbuf[xv + 3] ^ 0xAA;
		io1 = Si[o1];
		S1 = S[1];
		o2 = ivbuf[xv + 4] ^ 0xAA;
		io2 = Si[o2];
		S2 = S[2];
		Sq = S[q];
		dq = Sq + jj[q - 1];

		if (S2 == 0)
		{
			if ((S1 == 2) && (o1 == 2))
			{
				Kq = 1 - dq;
				votes[A_neg][Kq]++;
				Kq = 2 - dq;
				votes[A_neg][Kq]++;
			}
			else if (o2 == 0)
			{
				Kq = 2 - dq;
				votes[A_neg][Kq]++;
			}
		}
		else
		{
			if ((o2 == 0) && (Sq == 0))
			{
				Kq = 2 - dq;
				votes[A_u15][Kq]++;
			}
		}

		if ((S1 == 1) && (o1 == S2))
		{
			Kq = 1 - dq;
			votes[A_neg][Kq]++;
			Kq = 2 - dq;
			votes[A_neg][Kq]++;
		}

		if ((S1 == 0) && (S[0] == 1) && (o1 == 1))
		{
			Kq = 0 - dq;
			votes[A_neg][Kq]++;
			Kq = 1 - dq;
			votes[A_neg][Kq]++;
		}

		if (S1 == q)
		{
			if (o1 == q)
			{
				Kq = Si[0] - dq;
				votes[A_s13][Kq]++;
			}
			else if (((1 - q - o1) & 0xFF) == 0)
			{
				Kq = io1 - dq;
				votes[A_u13_1][Kq]++;
			}
			else if (io1 < q)
			{
				jq = Si[(io1 - q) & 0xFF];

				if (jq != 1)
				{
					Kq = jq - dq;
					votes[A_u5_1][Kq]++;
				}
			}
		}

		if ((io1 == 2) && (S[q] == 1))
		{
			Kq = 1 - dq;
			votes[A_u5_2][Kq]++;
		}

		if (S[q] == q)
		{
			if ((S1 == 0) && (o1 == q))
			{
				Kq = 1 - dq;
				votes[A_u13_2][Kq]++;
			}
			else if ((((1 - q - S1) & 0xFF) == 0) && (o1 == S1))
			{
				Kq = 1 - dq;
				votes[A_u13_3][Kq]++;
			}
			else if ((S1 >= ((-q) & 0xFF)) && (((q + S1 - io1) & 0xFF) == 0))
			{
				Kq = 1 - dq;
				votes[A_u5_3][Kq]++;
			}
		}

		if ((S1 < q) && (((S1 + S[S1] - q) & 0xFF) == 0) && (io1 != 1)
			&& (io1 != S[S1]))
		{
			Kq = io1 - dq;
			votes[A_s5_1][Kq]++;
		}

		if ((S1 > q) && (((S2 + S1 - q) & 0xFF) == 0))
		{
			if (o2 == S1)
			{
				jq = Si[(S1 - S2) & 0xFF];

				if ((jq != 1) && (jq != 2))
				{
					Kq = jq - dq;
					votes[A_s5_2][Kq]++;
				}
			}
			else if (o2 == ((2 - S2) & 0xFF))
			{
				jq = io2;

				if ((jq != 1) && (jq != 2))
				{
					Kq = jq - dq;
					votes[A_s5_3][Kq]++;
				}
			}
		}

		if ((S[1] != 2) && (S[2] != 0))
		{
			J2 = S[1] + S[2];

			if (J2 < q)
			{
				t2 = S[J2] + S[2];

				if ((t2 == q) && (io2 != 1) && (io2 != 2) && (io2 != J2))
				{
					Kq = io2 - dq;
					votes[A_s3][Kq]++;
				}
			}
		}

		if (S1 == 2)
		{
			if (q == 4)
			{
				if (o2 == 0)
				{
					Kq = Si[0] - dq;
					votes[A_4_s13][Kq]++;
				}
				else
				{
					if ((jj[1] == 2) && (io2 == 0))
					{
						Kq = Si[254] - dq;
						votes[A_4_u5_1][Kq]++;
					}
					if ((jj[1] == 2) && (io2 == 2))
					{
						Kq = Si[255] - dq;
						votes[A_4_u5_2][Kq]++;
					}
				}
			}
			else if ((q > 4) && ((S[4] + 2) == q) && (io2 != 1) && (io2 != 4))
			{
				Kq = io2 - dq;
				votes[A_u5_4][Kq]++;
			}
		}
	}
}

typedef struct
{
	int idx, val;
} vote;

static int cmp_votes(const void * bs1, const void * bs2)
{
	if (((vote *) bs1)->val < ((vote *) bs2)->val) return (1);

	if (((vote *) bs1)->val > ((vote *) bs2)->val) return (-1);

	return (0);
}

int main(int argc, char * argv[])
{
	FILE * f;
	long nb_ivs;
	int i, n, B, *vi;
	int votes[N_ATTACKS][256];

	unsigned char *ivbuf, *s;
	unsigned char buffer[4096];
	unsigned char wepkey[16];

	vote poll[64][256];

	if (argc != 3)
	{
		printf("usage: kstats <ivs file> <104-bit key>\n");
		return (1);
	}

	i = 0;
	s = (unsigned char *) argv[2];

	buffer[0] = s[0];
	buffer[1] = s[1];
	buffer[2] = '\0';

	while (sscanf((char *) buffer, "%x", &n) == 1)
	{
		if (n < 0 || n > 255)
		{
			fprintf(stderr, "Invalid wep key.\n");
			return (1);
		}

		wepkey[i++] = n;

		if (i >= 16) break;

		s += 2;

		if (s[0] == ':' || s[0] == '-') s++;

		if (s[0] == '\0' || s[1] == '\0') break;

		buffer[0] = s[0];
		buffer[1] = s[1];
	}

	if (i != 13)
	{
		fprintf(stderr, "Invalid wep key.\n");
		return (1);
	}

	if ((ivbuf = (unsigned char *) malloc(IVBUF_SIZE)) == NULL)
	{
		perror("malloc");
		return (1);
	}

	memset(ivbuf, 0, IVBUF_SIZE);

	if ((f = fopen(argv[1], "rb")) == NULL)
	{
		free(ivbuf);
		perror("fopen");
		return (1);
	}

	if (fread(buffer, 1, 4, f) != 4)
	{
		free(ivbuf);
		fclose(f);
		perror("fread header");
		return (1);
	}

	if (memcmp(buffer, "\xBF\xCA\x84\xD4", 4) != 0)
	{
		free(ivbuf);
		fclose(f);
		fprintf(stderr, "Not an .IVS file\n");
		return (1);
	}

	nb_ivs = 0;

	while (1)
	{
		if (fread(buffer, 1, 1, f) != 1) break;

		if (buffer[0] != 0xFF)
			if (fread(buffer + 1, 1, 5, f) != 5) break;

		if (fread(buffer, 1, 5, f) != 5) break;

		memcpy(ivbuf + nb_ivs * 5, buffer, 5);

		nb_ivs++;
	}

	for (B = 0; B < 13; B++)
	{
		for (i = 0; i < 256; i++)
		{
			poll[B][i].idx = i;
			poll[B][i].val = 0;
		}

		calc_votes(ivbuf, nb_ivs, wepkey, B, votes);

		for (n = 0, vi = (int *) votes; n < N_ATTACKS; n++)
			for (i = 0; i < 256; i++, vi++) poll[B][i].val += *vi * K_COEFF[n];

		qsort(poll[B], 256, sizeof(vote), cmp_votes);

		printf("KB %02d VALID  %02X", B, wepkey[B]);

		for (i = 0; i < 256; i++)
			if (poll[B][i].idx == wepkey[B]) printf("(%4d) ", poll[B][i].val);

		for (i = 0; i < N_ATTACKS; i++) printf("%3d  ", votes[i][wepkey[B]]);

		printf("\n");

		printf("KB %02d FIRST  %02X(%4d) ", B, poll[B][0].idx, poll[B][0].val);

		for (i = 0; i < N_ATTACKS; i++)
			printf("%3d  ", votes[i][poll[B][0].idx]);

		printf("\n");

		printf("KB %02d SECOND %02X(%4d) ", B, poll[B][1].idx, poll[B][1].val);

		for (i = 0; i < N_ATTACKS; i++)
			printf("%3d  ", votes[i][poll[B][1].idx]);

		printf("\n");

		printf("KB %02d THIRD  %02X(%4d) ", B, poll[B][2].idx, poll[B][2].val);

		for (i = 0; i < N_ATTACKS; i++)
			printf("%3d  ", votes[i][poll[B][2].idx]);

		printf("\n\n");
	}

	free(ivbuf);
	fclose(f);

	return (0);
}
